zerver/lib/webhooks/common.py

```diff
@@ -114,3 +130,104 @@ def validate_extract_webhook_http_header(request: HttpRequest, header: str,
+                A single-use function that will be called by passing this ThirdPartyAPIAmbassador
+                instance to it during initialization. There are some ready-made methods in the
+                Zulip library (in this same file) which handle basic schemes like sending an API
+                token with every request as a "GET" parameter.
+
+            request_preprocessor:
+                A function that should take this ThirdPartyAPIAmbassador instance and then do
+                some preprocessing, such as setting updating request_kwargs based on the previous
+                response from calling http_api_callback, starting a log entry, etc.
+                Argument(s) passed during the call: self
+
+            request_postprocessor:
+                Similar to the request_preprocessor, this should be a function that takes this
+                ThirdPartyAPIAmbassador object and then does some postprocessing such as logging
+                the results, triggering another subsequent call, PMing the owner of the bot, etc.
+                Argument(s) passed during the call: self
+
+            response_log: (internal use)
+                A list of all of the responses from calling http_api_callback (in chronological
+                order with index 0 being the oldest response and index -1 being the most recent).
+        """
+
+        if bot.is_bot is not True:
+            msg = "Ambassador must be a bot. {name} is not a bot".format(name=bot.full_name)
+            raise ValueError(msg)
+        self.bot = bot
+
+        self.request_kwargs = request_kwargs
+        self.request_preprocessor = request_preprocessor
+        self.request_postprocessor = request_postprocessor
+
+        self.response_log = []  # type: List[Response]
+
+        if authentication_handler:
+            authentication_handler(self)
+
+    @property
+    def result(self) -> Optional[Response]:
+        try:
+            return self.response_log[-1]
+        except IndexError:
+            return None
+
+    def http_api_callback(self, api_endpoint: str) -> Response:
+
+        if self.request_preprocessor:
+            self.request_preprocessor(self)
+
+        response = requests.get(api_endpoint, **self.request_kwargs)
+        self.response_log.append(response)
+
+        if self.request_postprocessor:
+            self.request_postprocessor(self)
+
+        return response
+
+def api_token_auth_handler(ambassador: ThirdPartyAPIAmbassador, get_param_key: str="token",
+                           config_element_key: str="api_token") -> None:
+    """ A readymade authentication handler method to handle the case where all we need to do is
+    add a simple "get" parameter to every URL we're contacting. This method throws a ValueError
+    if the configuration element does not exist. Thus, when using this as an authentication handler,
+    make sure that you wrap instantiation of the ThirdPartyAPIAmbassador in a try/except block. """
```

Comment:
As I stated above, I think docstrings such as this would be redundant once we have a working webhook that uses this code and serves as an example that we can use as a reference for future improvements.